README.md
# sdx-transform-cs

[![Build Status](https://github.com/ONSdigital/sdx-transform-cs/workflows/Build/badge.svg)](https://github.com/ONSdigital/sdx-transform-cs) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/0d8f1899b0054322b9d0ec8f2bd62d86)](https://www.codacy.com/app/ons-sdc/sdx-transform-cs?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=ONSdigital/sdx-transform-cs&amp;utm_campaign=Badge_Grade)

The sde-transform-cs app is used within the Office National of Statistics (ONS) for transforming Survey Data Exchange (SDX) Surveys to formats in use in Common Software.

## Prerequisites
The service has a dependency on the `pdf2ppm` commandline tool bundled in the poppler package. 
To check whether this is already installed, run:
```bash
$ make check-dependencies
```

To install `pdf2ppm` on a Mac, use:

```bash
$ brew install poppler
```

## Installation
This application presently installs required packages from requirements files:
- `requirements.txt`: packages for the application, with hashes for all packages: see https://pypi.org/project/hashin/
- `test-requirements.txt`: packages for testing and linting

It's also best to use `pyenv` and `pyenv-virtualenv`, to build in a virtual environment with the currently recommended version of Python.  To install these, see:
- https://github.com/pyenv/pyenv
- https://github.com/pyenv/pyenv-virtualenv
- (Note that the homebrew version of `pyenv` is easiest to install, but can lag behind the latest release of Python.)

### Getting started
Once your virtual environment is set, install the requirements:
```shell
$ make build
```

To test, first run `make build` as above, then run:
```shell
$ make test
```

NOTE: .pck and .nobatch test files are required to not have a newline character at the end of the file.
A simple way to remove it is to do the following command `perl -pi -e 'chomp if eof' filename`

It's also possible to build sdx-transform-cs within a container using docker. From the sdx-transform-cs directory:

```bash
$ docker build -t sdx-transform-cs .
```

## Usage

To start sdx-transform-cs service locally, use the following command:

```bash
$ python server.py
```

If you've built the image under docker, you can start using the following:

```bash
$ docker run -p 5000:5000 sdx-transform-cs
```

sdx-transform-cs by default binds to port 5000 on localhost. It exposes several endpoints for transforming to idbr and pck formats. It returns a response formatted in the type requested. Post requests are made aginst the uri endpoints /pck, /idbr, /images, /common-software or /cord. Responses are delivered in the format requested, except the /images, /common-software, /cord and /cora endpoints which return archived zips of requested data. There is also a health check endpoint (get /healtcheck), which returns a json response with a key/value pairs describing the service state.

### Example

The example below uses the Python library [requests](https://github.com/kennethreitz/requests) to confirm some data is valid using sdx-transform-cs.

```python
import requests

data_to_transform = '''{
   "type": "uk.gov.ons.edc.eq:surveyresponse",
   "origin": "uk.gov.ons.edc.eq",
   "survey_id": "023",
   "version": "0.0.1",
   "collection": {
     "exercise_sid": "hfjdskf",
     "instrument_id": "0203",
     "period": "0216"
   },
   "submitted_at": "2016-03-12T10:39:40Z",
   "metadata": {
     "user_id": "789473423",
     "ru_ref": "12345678901A"
   },
   "data": {
     "11": "01/04/2016",
     "12": "31/10/2016",
     "20": "1800000",
     "51": "84",
     "52": "10",
     "53": "73",
     "54": "24",
     "50": "205",
     "22": "705000",
     "23": "900",
     "24": "74",
     "25": "50",
     "26": "100",
     "21": "60000",
     "27": "7400",
     "146": "some comment"
   }
}'''

r = requests.post('http://127.0.0.1:5000/pck', data=data_to_transform)

r.data =

'''FBFV03000012/03/16
FV          
RSI7B:12345678901A:0216'''
```

## Configuration

Some of important environment variables available for configuration are listed below:

| Environment Variable    | Default                               | Description
|-------------------------|---------------------------------------|----------------
| SDX_SEQUENCE_URL        | `http://sdx-sequence:5000`            | URL of the ``sdx-sequence`` service
| FTP_PATH                | `\\`                                  | FTP path
| SDX_FTP_IMAGE_PATH      | `EDC_QImages`                         | Location of EDC Images

## Image generation

When hitting certain endpoints, a zip file that contains a number of JPG images that represent the submission will be returned.
These images are produced with the help of JSON files that describe what the image should look like.  These can be found in
`transform/surveys/*.json`

The keys of these json files describe how the image should look and below is a guide on the what they do.

- `title`: Full survey name, appears at the top as a header
- `survey_id` : Id of the survey, appears below the title
- `form_type` : Form type of the survey,
- `question_groups`:  An array of sections.  Each section is comprised of a heading
and a number of questions relating to that section.

Each element of the question_groups is made up of the following:

- `title`:  Name of the section, appears at the top of the section
- `questions`: Array of questions.  Each question corresponds to a question the respondent would've filled out on EQ.

Each element of questions is made up of the following:

- `text`: This is the heading of the question.  This should be as close as possible to the question asked in EQ.
- `question_id`: This is the qcode of the question.  The code in the data from EQ will be used to populate the answer of this field
- `number`: Used to output the qcode of the question in the image.
- `type`: Can be one of `currency`, `date`, `checkbox`, `radio`, `contains`, `positiveinteger` or `percentage`. None of these affect the image.
The `contains` and `date` types affect the data added to the pck file.  The other types are there to aid in maintaining the survey.
- `options`:  Doesn't affect the image.  It's used to add context to radio and checkbox fields as each possible answer will have its own qcode
but needs to have the same question because of the way it needs to look in the image.

## License

Copyright © 2016, Office for National Statistics (https://www.ons.gov.uk)

Released under MIT license, see [LICENSE](LICENSE) for details.


CHANGELOG.md
### 4.5.0 2021-06-29
 - Construction changes

### 4.4.0 2021-02-22
 - ukis survey update for 2020
 - Renaming of all file outputs to use tx_id
 - Low Carbon survey onboarded
    - JSON files added
    - transformer added  

### 4.3.2 2020-11-12
- Update MES qcodes 1086 and 1087 to round to the nearest thousand

### 4.3.1 2020-11-06
 - Fix broken transformation in MES due to wording changes
 - Added missing commas from mes_transform

### 4.3.0 2020-10-30
 - Fix for MBS comments

### 4.2.0 2020-10-30
 - Changes to MBS comments
 - Changes to MES

### 4.1.0 2020-10-20
 - Added MES transformer

### 4.0.0 2020-09-07
 - Major refactor
 - Add EPE transformer
 - Moved logging setup

### 3.21.0 2020-08-05
 - Fix E-commerce index file period 

### 3.20.0 2020-07-28
 - Ensure IDBR receipts have a 6 digit period

### 3.19.0 2020-07-07
 - Remove Vacancies formtype mapping to 181 for indices  
 - Add openapi.yml
 - Add uml diagrams

### 3.18.0 2020-06-09
 - Add construction formtype 0002
 - Remove Cloudfoundry deployment files

### 3.17.1 2020-05-21
 - Updated packages

### 3.17.0 2020-04-27
 - Fix construction transformation when total value is zero and missing qcodes 902-904 

### 3.16.0 2020-04-17
 - Add construction survey

### 3.15.0 2020-03-20
 - Handle new comment field in QBS (139)

### 3.14.1 2020-03-03
 - Added implicit percentages for a number of sets of questions in Ecommerce (0001, 0002)
 - Changed filename generation in vacancies to match the downstream requirements (IDBR unchanged, everything else changed to survey id 181)
 - Fixed the titles for Vacancies images.

### 3.14.0 2020-01-30
 - Add Vacancies survey

### 3.13.1 2019-12-17
 - Fix transformation errors in e-commerce 0001 and 0002 formtypes

### 3.13.0 2019-11-25
 - Add E-commerce 2019 survey
 - Update packages and fix security issue in Pillow

### 3.12.3 2019-11-08
 - Add python 3.8 to travis builds

### 3.12.2 2019-09-26
 - Update werkzeug to 0.15.6 to fix security issue
 - Update various packages to keep them current

### 3.12.1 2019-09-04
 - Update documentation
 - Remove unused `comments` value in JSON files used for image generation

### 3.12.0 2019-08-13
 - Remove unused `/html` and `/pdf` endpoints

### 3.11.2 2019-06-28
 - Fix incorrect value in stocks formtypes 0001 and 0002

### 3.11.1 2019-06-19
 - Update structlog package version
 - Fix incorrect value in stocks formtype 0061

### 3.11.0 2019-05-31
 - Renamed QSI (Quarterly Stocks Inquiry) to QSS (Quarterly Stocks Survey)
 - Add calculated totals for QSS as the totals aren't passed to SDX
 - Fix missing date qcodes (11 and 12) for QSS (Stocks)
 - Add formtype 0070 for QSS (Stocks)
 - Change rounding to stocks to divide by 1000 after rounding (like QCAS)

### 3.10.0 2019-04-02
 - Update Jinja to 2.10.1 to fix a security issue.
 - Update Pillow and PyYAML versions
 - Updated the dockerfile
 - Used Python 3.5 inbuilt dictionary merging instead of merge_dicts
 - Added support for QSI (Stocks) survey

### 3.9.0 2019-02-19
 - Added support for QPSES survey

### 3.8.0 2019-02-08
 - Added support for UKIS and CORA surveys

### 3.7.1 2018-12-20
 - Change option value for E-Commerce q-code 416 (r3)
 - Ensure r3 is on image file

### 3.7.0 2018-12-12
  - Added support for E-commerce form and CORD surveys.
  - Improve tests for mbs_transformer

### 3.6.0 2018-09-11
  - Added support for QCAS form types.
  - Calculate total values for QCAS to be sent downstream.
  - Replace all negative numbers with a string of 11 characters containing only 9's in pck file.
    
### 3.5.2 2018-08-09
  - Fix error in turnover rounding

### 3.5.1 2018-07-12
  - Add turnover question to 009.0167 (MBS)

### 3.5.0 2018-06-28
  - Change mbs transformer q_code from d40 to d49

### 3.4.0 2018-05-21
  - Add MBS Transformer

### 3.3.0 2018-03-26
  - Impute zero values for MWSS gross pay breakdown questions when the respondent has confirmed they are zero

### 3.2.0 2018-03-22
  - Impute zero values for RSI and QBS total questions when the respondent has confirmed they are zero

### 3.1.0 2018-02-20
  - Changes to the MWSS survey.  Question text and format changes in the schema. Also, questions beginning 90 and 190 aggregate to 90 in the downstream file.  Questions 91-97 and 191-197 no longer need to be included in the aggregation.

### 3.0.0 2018-01-04
  - Added the original Json data in the zip file sent to the ftp server
  - refactored cs_fromatter to use named parameters instead of **kwargs for clarity
  - Add /info healthcheck endpoint
  - Remove reference to ftp server

### 2.13.0 2017-11-27
  - Bug fix - nothing displayed after the decimal point in decimal responses in image files
  - Move transformers to be memory based instead of file based
  - Initial files for move to cloudfoundry

### 2.12.0 2017-11-09
  - Fix incorrect index date time

### 2.11.0 2017-11-01
  - Removed unchanging configurable variables.
  - Add all service configurations to config file
  - Correct the spelling error in RSI image
  - Rename FTP_HOST to FTP_PATH
  - Improve test output and implemented Pytest

### 2.10.0 2017-10-18
  - Refactored variable names for MWSS transformer

### 2.9.0 2017-10-16
  - Support decimal values in MWSS Surveys ( survey_id 005 )

### 2.8.0 2017-10-02
  - Round decimal values in RSI surveys ( survey_id 023 )
  - Update deleting tmp files

### 2.7.0 2017-09-25
  - Removed SDX common clone in docker
  - Restore defaults for qids 130, 131, 132.
  - Remove pip `--require-hashes` requirement
  - Make use of sequence list endpoint
  - Populate default dates on certain surveys from metadata
  - Add fix for cryptography install in docker

### 2.6.0 2017-08-23
  - Update sdx-common version
  - Ensure integrity and version of library dependencies

### 2.5.0 2017-07-25
  - EQ MCI Survey Refresh
  - Change all instances of ADD to COPY in Dockerfile

### 2.4.0 2017-07-10
  - Update and pin version of sdx-common to 0.7.0
  - Remove use of SDX_HOME in makefile

### 2.3.0 2017-06-19
  - Timestamp all logs as UTC

transform-pck.py
from jinja2 import Environment, PackageLoader
from datetime import datetime

import dateutil.parser
env = Environment(loader=PackageLoader('transform', 'templates'))

form_ids = {
  "0203": "RSI7B",
  "0205": "RSI9B",
  "0213": "RSI8B",
  "0215": "RSI10B"
}

form_questions = {
  "0203": [1, 11, 12, 20, 21, 22, 23, 24, 25, 26, 146],
  "0205": [1, 11, 12, 20, 21, 22, 23, 24, 25, 26, 27, 146],
  "0213": [1, 11, 12, 20, 21, 22, 23, 24, 25, 26, 50, 51, 52, 53, 54, 146],
  "0215": [1, 11, 12, 20, 21, 22, 23, 24, 25, 26, 27, 50, 51, 52, 53, 54, 146]
}

form_question_types = {
  "1": 'contains',
  "11": 'date',
  "12": 'date',
  "146": 'contains'
}

response = {
   "type": "uk.gov.ons.edc.eq:surveyresponse",
   "version": "0.0.1",
   "origin": "uk.gov.ons.edc.eq",
   "survey_id": "023",
   "collection": {
     "exercise_sid": "hfjdskf",
     "instrument_id": "0203",
     "period": "0216"
   },
   "submitted_at": "2016-03-12T10:39:40Z",
   "metadata": {
     "user_id": "789473423",
     "ru_ref": "12345678901A"
   },
   "data": {
     "11": "01/04/2016",
     "12": "31/10/2016",
     "20": "1800000",
     "51": "84",
     "52": "10",
     "53": "73",
     "54": "24",
     "50": "205",
     "22": "705000",
     "23": "900",
     "24": "74",
     "25": "50",
     "26": "100",
     "21": "60000",
     "27": "7400",
     "146": "some comment"
   }
}


def get_derived_value(question_id, value):
    if question_id in form_question_types:
        form_question_type = form_question_types[question_id]

        if form_question_type == 'contains':
            value = "1" if value else "2"
        elif form_question_type == 'date':
            derived_date = datetime.strptime(value, "%d/%m/%Y")

            value = derived_date.strftime("%d%m%y")

    return value.zfill(11)


def derive_answers(answers, instrument_id):
    answers = []

    for k, v in response['data'].items():
        if int(k) in form_questions[instrument_id]:
            answers.append((int(k), get_derived_value(k, v)))

    return sorted(answers)


template = env.get_template('pck.tmpl')

instrument_id = response['collection']['instrument_id']

submission_date = dateutil.parser.parse(response['submitted_at'])
submission_date_str = submission_date.strftime("%d/%m/%y")

form_id = form_ids[instrument_id]

answers = derive_answers(response['data'].items(), instrument_id)

print(template.render(response=response, submission_date=submission_date_str, batch_number=30000, form_id=form_id, answers=answers))
